src/main/java/com/google/api/gax/grpc/ServiceApiSettings.java

```diff
@@ -66,34 +98,51 @@
   /**
    * Constructs an instance of ServiceApiSettings.
    */
-  protected ServiceApiSettings(ManagedChannel channel,
-                               boolean shouldAutoCloseChannel,
-                               ScheduledExecutorService executor,
-                               ConnectionSettings connectionSettings,
-                               String generatorName,
-                               String generatorVersion,
-                               String clientLibName,
-                               String clientLibVersion) {
-    this.channel = channel;
-    this.executor = executor;
-    this.connectionSettings = connectionSettings;
-    this.shouldAutoCloseChannel = shouldAutoCloseChannel;
+  protected ServiceApiSettings(
+      ChannelProvider channelProvider,
+      ExecutorProvider executorProvider,
+      String generatorName,
+      String generatorVersion,
+      String clientLibName,
+      String clientLibVersion) {
+    this.channelProvider = channelProvider;
+    this.executorProvider = executorProvider;
     this.clientLibName = clientLibName;
     this.clientLibVersion = clientLibVersion;
     this.generatorName = generatorName;
     this.generatorVersion = generatorVersion;
   }
 
-  public final ManagedChannel getChannel() {
-    return channel;
+  /**
+   * Return the channel to be used to connect to the service, retrieved using the channelProvider.
+   * If no channel was set, a default channel will be instantiated.
+   */
+  public final ManagedChannel getOrBuildChannel() throws IOException {
+    return getChannelProvider().getChannel(getOrBuildExecutor());
```

Comment:
Yes good point. Can implement by changing getOrBuildChannel to buildChannel, calling once in XApi constructor and passing channel directly to ApiCallable.create instead of passing Settings object. To prevent channel re-use, perhaps when a ChannelProvider has a fixed channel, it should throw if buildChannel is called more than once? Also, does the same thing make sense for ExecutorProvider, or is it sensible to reuse an Executor between instances of XApi?